platform/admin/core/admin-core-appservice/src/main/java/org/codice/ddf/admin/application/service/migratable/ProfileMigratable.java

```diff
@@ -0,0 +1,280 @@
+
+  public static final String PROFILE_VERSION = "version";
+
+  public static final String PROFILE_DESCRIPTION = "description";
+
+  public static final String PROFILE_REPOSITORY = "repository";
+
+  public static final String PROFILE_URI = "uri";
+
+  public static final String PROFILE_LOCATION = "location";
+
+  public static final String PROFILE_STATE = "state";
+
+  public static final String PROFILE_START = "start";
+
+  public static final String PROFILE_REGION = "region";
+
+  public static final String PROFILE_APPS = "apps";
+
+  public static final String PROFILE_FEATURES = "features";
+
+  public static final String PROFILE_BUNDLES = "bundles";
+
+  /**
+   * Holds a retry count.
+   *
+   * <p>Because it was seen with Karaf that sometimes installing or starting a feature or a bundle
+   * might have side effects and install or start others which were not supposed to in the original
+   * system. As such, we are going to attempt multiple times for each operations such that we can
+   * account for side effects of other operations. In addition, because we are unable to determine a
+   * proper order to process all features or bundles, it is possible that we end up attempting to
+   * install a feature for which the dependencies have not yet been installed. We are therefore
+   * forced to attempt multiple times hoping that dependencies are properly being handled by the
+   * previous pass. Failures will be reported only on the last attempt. Retries will automatically
+   * stop when all apps, features, and apps are determined to be as they were at export time.
+   */
+  @SuppressWarnings("PMD.DefaultPackage" /* designed to be used by TaskList within this package */)
+  static final int RETRY_COUNT = 5;
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(ProfileMigratable.class);
+
+  /**
+   * Holds the current export version.
+   *
+   * <p>1.0 - initial version
+   */
+  private static final String VERSION = "1.0";
+
+  private static final Path PROFILE_PATH = Paths.get("profile.json");
+
+  @VisibleForTesting static final ObjectMapper MAPPER = JsonFactory.create();
+
+  private final ApplicationManager appMgr;
+
+  private final FeatureManager featureMgr;
+
+  private final BundleManager bundleMgr;
+
+  @VisibleForTesting
+  ProfileMigratable(ApplicationManager appMgr, FeatureManager featureMgr, BundleManager bundleMgr) {
+    this.appMgr = appMgr;
+    this.featureMgr = featureMgr;
+    this.bundleMgr = bundleMgr;
+  }
+
+  /**
+   * Instantiates a profile migratable with the specified application and features services.
+   *
+   * @param appService the application services to use
+   * @param featuresService the features service to use
+   * @throws IllegalArgumentException if <code>applicationService</code> or <code>featuresService
+   * </code> is <code>null</code>
+   */
+  public ProfileMigratable(ApplicationService appService, FeaturesService featuresService) {
```

Comment:
❓ Any reason we're not just creating the 3 "managers" in the blueprint file and injecting them? That would eliminate the need for two constructors.

Reply:
Never thought of doing that, I guess I am just not used to "injection-based" configuration.